java/core/src/main/java/com/google/protobuf/DescriptorMessageInfoFactory.java

```diff
@@ -593,21 +616,92 @@ static String getFieldName(FieldDescriptor fd) {
     String name = (fd.getType() == FieldDescriptor.Type.GROUP)
                   ? fd.getMessageType().getName()
                   : fd.getName();
-    String suffix = specialFieldNames.contains(name) ? "__" : "_";
-    return snakeCaseToCamelCase(name) + suffix;
+
+    // convert to UpperCamelCase for comparison to the specialFieldNames
+    // (which are in UpperCamelCase)
+    String upperCamelCaseName = snakeCaseToUpperCamelCase(name);
+
+    String suffix = specialFieldNames.contains(upperCamelCaseName)
+                  // For field names that match the specialFieldNames,
+                  // append "__" to prevent field accessor method names from
+                  // clashing with other methods.
+                  ? "__"
+                  // For other field names, append "_" to prevent field names
+                  // from clashing with java keywords.
+                  : "_";
+    return snakeCaseToLowerCamelCase(name) + suffix;
   }
 
   private static String getCachedSizeFieldName(FieldDescriptor fd) {
-    return snakeCaseToCamelCase(fd.getName()) + "MemoizedSerializedSize";
+    return snakeCaseToLowerCamelCase(fd.getName()) + "MemoizedSerializedSize";
+  }
+
+  /**
+   * Converts a snake case string into lower camel case.
+   *
+   * <p>
+   * Some examples:
+   * <pre>
+   *     snakeCaseToLowerCamelCase("foo_bar") => "fooBar"
+   *     snakeCaseToLowerCamelCase("foo") => "foo"
+   * </pre>
+   * </p>
+   *
+   * @param snakeCase the string in snake case to convert
+   * @return the string converted to camel case, with a lowercase first character
+   */
+  private static String snakeCaseToLowerCamelCase(String snakeCase) {
+  	return snakeCaseToCamelCase(snakeCase, false);
   }
 
   /**
-   * This method must match exactly with the corresponding function in protocol compiler. See:
-   * https://github.com/google/protobuf/blob/v3.0.0/src/google/protobuf/compiler/java/java_helpers.cc#L153
+   * Converts a snake case string into upper camel case.
+   *
+   * <p>
+   * Some examples:
+   * <pre>
```

Comment:
pre should be outside the p

Reply:
why would you put \<pre> outside the \<p>   everything inside \<pre> is pre rendered stuff (which \<p> is not)

Reply:
ah, I think I understand what you meant now.  (I thought you meant nest the p inside the pre, but you just meant separate the two).
Fixed